infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -0,0 +1,73 @@
+package org.corfudb.infrastructure.orchestrator.actions;
+
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.infrastructure.orchestrator.Action;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.OutrankedException;
+import org.corfudb.runtime.view.Layout;
+
+/**
+ * This action attempts to restore redundancy for all servers across all segments
+ * starting from the oldest segment. It then collapses the segments once the set of
+ * servers in the 2 oldest subsequent segments are equal.
+ * Created by Zeeshan on 2019-02-06.
+ */
+public class RestoreRedundancyMergeSegments extends Action {
+
+    /**
+     * Returns set of nodes which are present in the specified index but not in the previous segment. These
+     * nodes have reduced redundancy and state needs to be transferred only to these before these segments can be
+     * merged.
+     *
+     * @param layout             Current layout.
+     * @param layoutSegmentIndex Segment to compare to get nodes with reduced redundancy.
+     * @return Set of nodes with reduced redundancy.
+     */
+    private Set<String> getNodesWithReducedRedundancy(Layout layout, int layoutSegmentIndex) {
+        // Get the set of servers present in the second segment but not in the first
+        // segment.
+        return Sets.difference(
+                layout.getSegments().get(layoutSegmentIndex).getAllLogServers(),
+                layout.getSegments().get(layoutSegmentIndex - 1).getAllLogServers());
+    }
+
+    @Nonnull
+    @Override
+    public String getName() {
+        return "RestoreRedundancyAndMergeSegments";
+    }
+
+    @Override
+    public void impl(@Nonnull CorfuRuntime runtime) throws OutrankedException,
+            ExecutionException,
+            InterruptedException {
+
+        // Refresh layout.
+        runtime.invalidateLayout();
+        Layout layout = runtime.getLayoutView().getLayout();
+
+        // Catchup all servers across all segments.
+        while (layout.getSegments().size() > 1) {
+
+            Set<String> lowRedundancyServers = getNodesWithReducedRedundancy(layout, 1);
+
+            // Currently the state is transferred for the complete segment.
+            // TODO: Add stripe specific transfer granularity for optimization.
+            // Transfer the replicated segment to the difference set calculated above.
+            StateTransfer.transfer(layout, lowRedundancyServers, runtime, layout.getSegments().get(0));
+
```

Comment:
can you please provide `getFirstSegment()` method instead of `getSegments().get(0)`

Reply:
Done.